precompilepkgs: make the circular dep warning clearer and more informative

IanButterworth opened the pull request:
Was e.g.
```
┌ Warning: Circular dependency detected. Precompilation will be skipped for:
│   Base.PkgId(Base.UUID("eb0c05c4-6780-5852-a67e-5d31d2970b9a"), "ArrayInterfaceTrackerExt")
│   Base.PkgId(Base.UUID("f517fe37-dbe3-4b94-8317-1923a5111588"), "Polyester")
│   Base.PkgId(Base.UUID("0d7ed370-da01-4f52-bd93-41d350b8b718"), "StaticArrayInterface")
│   Base.PkgId(Base.UUID("6a4ca0a5-0e36-4168-a932-d9be78d558f1"), "AcceleratedKernels")
│   Base.PkgId(Base.UUID("244f68ed-b92b-5712-87ae-6c617c41e16a"), "NNlibAMDGPUExt")
│   Base.PkgId(Base.UUID("06b0261c-7a9b-5753-9bdf-fd6840237b4a"), "StaticArrayInterfaceStaticArraysExt")
│   Base.PkgId(Base.UUID("21141c5a-9bdb-4563-92ae-f87d6854732e"), "AMDGPU")
│   Base.PkgId(Base.UUID("9f7883ad-71c0-57eb-9f7f-b5c9e6d3789c"), "Tracker")
└ @ Base.Precompilation precompilation.jl:511
```
Now
![Screenshot 2024-11-21 at 11 20 50 PM](https://github.com/user-attachments/assets/6939d834-90c3-4d87-baa9-cf6a4931ca03)

Thanks to @topolarity figuring out proper cycles tracking.

IanButterworth committed:
make the circular dep report prettier

topolarity commented:
Is it possible to print the cycle as a chain?

It's not clear to me whether or not I can just read this list from top to bottom to get the cycle chain, so making that more obvious seems helpful

IanButterworth commented:
I don't immediately know if these are ordered meaningfully. You might be better suited to answer that.. 

Also, what if there are multiple circularities.

IanButterworth committed:
show proper cycle information

Co-Authored-By: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

IanButterworth committed:
make arrows clearer

IanButterworth committed:
Revert "make arrows clearer"

This reverts commit 274608237ec7f75d83fee8bc7a46364e14e178ab.

topolarity committed:
Bring back `could_be_cycle` for efficiency

This is an important optimization for us not to lose, since we scan the
deps graph very redundantly.

topolarity commented:
We should probably backport to 1.10 too, since we expect that users will be hitting extension cycles a _lot_ more often on that release.

In 1.11 these cycles are supposed to be impossible, except for environment tearing (where you end up with a package graph that doesn't actually `] resolve`) and any implementation bugs (some of which were not fixed until.. yesterday 😉)

Still very good to have (esp. on 1.10), just want to be clear about whether this is a "normal" error any more

IanButterworth committed:
show cycles as indented

topolarity commented on base/precompilation.jl:
```diff
@@ -564,7 +591,26 @@ function _precompilepkgs(pkgs::Vector{String},
+                cycle_str *= line
+            end
+            cycles_names *= cycle_str
+        end
```
Making a note here, so this isn't forgotten (mostly by me):

This also needs to display any packages in `circular_deps` that weren't in any cycle (i.e. they depend on a dependency cycle, so they were excluded from pre-compile)

IanButterworth replied:
I just updated it to show these. Example up top

IanButterworth committed:
also list those outside the cycles & formatting

topolarity commented on base/precompilation.jl:
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
+        Precompilation will be skipped for dependencies all in $plural1:
+        $cycles_names"""
+        if !isempty(outer_deps)
+            msg *= "Precompilation will also be skipped for the following, which have depdenencies in the $plural2:\n"
```
```suggestion
            msg *= "Precompilation will also be skipped for the following, which have dependencies in the $plural2:\n"
```

topolarity commented on base/precompilation.jl:
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
+        plural1 = length(cycles) > 1 ? "these cycles" : "this cycle"
+        plural2 = length(cycles) > 1 ? "cycles" : "cycle"
+        msg = """Circular dependency detected.
+        Precompilation will be skipped for dependencies all in $plural1:
```
```suggestion
        Precompilation will be skipped for dependencies in $plural1:
```

topolarity commented on base/precompilation.jl:
```diff
@@ -564,7 +591,35 @@ function _precompilepkgs(pkgs::Vector{String},
     end
     if !isempty(circular_deps)
-        @warn """Circular dependency detected. Precompilation will be skipped for:\n  $(join(string.(circular_deps), "\n  "))"""
+        outer_deps = copy(circular_deps)
```
This warning message should probably be outlined to a separate function

IanButterworth committed:
another plural tweak

topolarity commented on base/precompilation.jl:
```diff
@@ -564,7 +591,36 @@ function _precompilepkgs(pkgs::Vector{String},
+        Precompilation will be skipped for dependencies all in $plural1:
+        $cycles_names"""
+        if !isempty(outer_deps)
+            msg *= "Precompilation will also be skipped for the following, which $plural2 depdenencies in the $plural3:\n"
```
```suggestion
            msg *= "Precompilation will also be skipped for the following, which depend on the above $plural3:\n"
```

I find this wording clearer, but the original is fine too - just make sure to fix the spelling of `dependencies`

topolarity reviewed (approved):
Looking great! I think it's much clearer to users what's going on here now.

Thanks for iterating on this @IanButterworth

topolarity commented:
Oh the cycle order does look a bit "backwards" to me

I would have expected the dependency of a package to be above it, not below

IanButterworth committed:
Apply suggestions from code review

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

IanButterworth committed:
separate to a function

IanButterworth commented:
> I would have expected the dependency of a package to be above it, not below

I think of it the other way. Matching the sentence order for "Foo depends on Bar, which depends on Baz"
```
Foo
 Bar
  Baz
```

topolarity commented:
Hmm.. What if we reverse the indentation from greatest to smallest?
```
  Baz
 Bar
Foo
```

Does that align our intuitions?

topolarity committed:
Fix-up extracted function

topolarity committed:
Fix printing for 1-element cycles

IanButterworth commented on base/precompilation.jl:
```diff
@@ -566,33 +610,52 @@ function _precompilepkgs(pkgs::Vector{String},
     for pkg in keys(direct_deps)
-        if scan_pkg!(pkg, direct_deps)
+        @assert isempty(stack)
+	if scan_pkg!(pkg, direct_deps)
```
```suggestion
        if scan_pkg!(pkg, direct_deps)
```

IanButterworth committed:
fix whitespace

topolarity reviewed (approved):
I still find the staircase weird here (and my first instinct is to read it "backwards"), but this is a big improvement and all the functionality looks ship-shape to me 

Good to merge, I think!

topolarity commented:
1.10 backport is done, but we should follow-up with 1.11 soon
